### ISSUE TITLE
Speed up discovery time

### ISSUE THREAD

**Issue by frblondin** [xunit/xunit#964]
I'm experiencing really slow unit test discovery time in Visual Studio. I did a very simple test that uses a T4 template to create thousands of tests (5k). I created three different assemblies: one using MSTest, on using MSTest2 (that requires another test adapter), and of course another one using Xunit2. Result: 6 seconds for the first two assemblies, and almost 3 minutes for the assembly using XUnit2.

Then I profiled the discovery and saw that 97.05% of the time was used by _TransparentProxyStub_Context@0. By looking at the callers & callees I could see that this overhead is coming from the ITestCase instances being shared between the AppDomain (one per assembly being analyzed) and the main AppDomain. ITestCase instances are of type TestMethodTestCase which is not being serialized but marshaled by ref, hence the issue observed above.

I checked how this is being implemented in https://www.nuget.org/packages/MSTest.TestAdapter/ and everything is almost the same, except for the instances being shared that are serializable instead of being marshaled.

To make sure that the diagnostic was the good one I enforced VsTestRunner.AppDomainDefaultBehavior to AppDomainSupport.Denied and the time was comparable to what I'm getting with MSTest 1 or 2. This cannot be used as a good fix since using the main AppDomain prevents from releasing the lock so that the assembly can no longer be compiled unless I kill vstest.discoveryengine.x86.exe.

So... here are my questions:
- Why isn't the TestMethodTestCase type and the other dependencies marked as serializable?
- If the TestMethodTestCase is used in multiple places and if making it serializable is complex, couldn't we introduce a simplified interface & implementation that only contains what is required by the discovery so that it would be easily made serializable? (it only needs the class name, method name, uniqueID, display name, the trait thunk, path of file, and line number)

From what I saw on forums it seems to be an issue for many people. At my work we have 3k+ tests and it takes more than 2 minutes to discover them all.

**Comment by bradwilson** [comment]
Duplicate of #472, which is already fixed (and will be shipped with 2.2 beta 3).

**Comment by frblondin** [comment]
Great, good news!

**Comment by frblondin** [comment]
Last question - there is no due date in the [milestone](https://github.com/xunit/xunit/milestone/25)... any idea?

**Comment by bradwilson** [comment]
You can use a build from our [CI feed on MyGet](http://myget.org/gallery/xunit) today. We are currently waiting for preview 3 of the .NET Core tooling, which does not yet have a public release date.

**Comment by frblondin** [comment]
I just tested this new version and I'm sorry to tell you that the time is still really slow...

As long as the test cases and marshaled by ref there will still be this issue.

**Comment by bradwilson** [comment]
Did you update the runners? The performance improvements are there, not in the core framework itself.

**Comment by frblondin** [comment]
Yep, I updated the runner (added the myget feed, removed the older runner, and added the preview 3).

Do you mind reopening the request and removing the duplicate label?

Also, I'd be interested in working with you to fix this and it would be helpful first to answer my two questions in the original comment.

Thx

**Comment by bradwilson** [comment]
> Why isn't the TestMethodTestCase type and the other dependencies marked as serializable?

They contain information which is only valid in the execution app domain (references to `Type`, `MethodInfo`, etc., from the unit test DLL).

> If the TestMethodTestCase is used in multiple places and if making it serializable is complex, couldn't we introduce a simplified interface & implementation that only contains what is required by the discovery so that it would be easily made serializable?

Test cases already support our custom serialization system (we cannot use any of the built-in serialization systems because they don't exist for PCLs). See: https://github.com/xunit/xunit/blob/e84a6208745312b726ab6080921837b78984df43/src/xunit.execution/Sdk/Frameworks/TestMethodTestCase.cs#L256-L269

Note that this serialization system solely exists for the Visual Studio runner, because it performs discovery and execution in separate processes. The purpose is not to shuttle them between the runner and execution app domains; rather, it is to serialize them in one execution app domain (the discovery process) so that they can be re-hydrated in another execution app domain (the execution process). These objects are never intended to live inside the runner app domain, but rather be called by remoting (since everything they reference, and everything they do, belongs to the execution app domain).

**Comment by frblondin** [comment]
I believe that ITestCase is used in two different cases:
- Test discovery & execution in the same process: we need all information available. Note that if these objects are shared among AppDomains I think it would still be more performant to share a serializable instance that contains only serializable data. When the AppDomain that later runs the test case and needs the method info / Type / ... we can use reflection to get back these objects. But in any

**Comment by frblondin** [comment]
Even if Visual Studio uses two different processes a profiling can easily prove that sharing by ref marshaled objects between AppDomain is a cause of performance problems, especially since many properties are accessed from outside the instance belongs to.

![xunitdiscoveryprofiling](https://cloud.githubusercontent.com/assets/5735134/18291299/5786152e-7487-11e6-961c-d666361d88bd.png)

I think that XUnit should manage two different interfaces for test cases:
1. Discovery test case that contains really basic information that can be serialized - no need to use MarshalByRef,
2. Existing test case which is only required to be instantiated in the AppDomain that will execute the test.

I made a proposal of what could be changed in xunit.abstractions before doing other changes in the other assemblies. See https://github.com/frblondin/xunit/commit/15616de72a95c40c1ff659a40dd9af914fda8f54

The idea is that the main app domain would just use ITestCaseDiscovery instances. When the test case needs to be executed we would provide the SerializedData property value so that the other AppDomain / process will be able to deserialize the ITestCase instance. This will add a call to deserialize but I can bet that this overhead will be negligible compared to the benefit of not calling so many marshaled instances anymore.

What do you think?

**Comment by bradwilson** [comment]
Changing abstractions would be a major version change, because it breaks back/forward compatibility. At this moment in time, we have no plans (ever!) to ship a 3.0.

**Comment by frblondin** [comment]
>> Why isn't the TestMethodTestCase type and the other dependencies marked as serializable?

> They contain information which is only valid in the execution app domain (references to Type, MethodInfo, etc., from the unit test DLL).

What is nice is that XUnit is abstracting the reflection types: the interfaces (ITypeInfo, IMethodInfo...) are not creating a dependency with System.Reflection types.

No?

**Comment by frblondin** [comment]
The root problem is that we have to create separate AppDomain, even for discovery, just because the .Net reflection always lock files. I even tried another approach that uses Assembly.ReflectionOnlyLoad(File.ReadAllBytes(...)): the assembly is no longer locked but the assembly can be loaded only once.

On the other hand there are other ways to get information from assemblies without using System.Reflection.

I did the simple test below using Mono.Cecil. Have you ever considered using another way to perform reflection for discovery? Reflection could be done in the main domain...

```
    class Program
    {
        private static string _directory;

        static void Main(string[] args)
        {
            var file = @"somePathToALibrary.dll";
            _directory = Path.GetDirectoryName(file);
            AppDomain.CurrentDomain.ReflectionOnlyAssemblyResolve += ReflectionOnlyAssemblyResolve;
            while (true)
            {
                var assembly = AssemblyDefinition.ReadAssembly(file);
                var methods = (from module in assembly.Modules
                               from t in module.Types
                               from m in t.Methods
                               where m.CustomAttributes.Any(a => a.AttributeType.FullName.Equals(typeof(FactAttribute).FullName))
                               select m).ToList();
                Console.WriteLine($"{methods.Count} unit tests resolved. Once more?");
                Console.ReadLine();
            }
        }

        private static Assembly ReflectionOnlyAssemblyResolve(object sender, ResolveEventArgs args)
        {
            var assemblyName = new AssemblyName(args.Name);
            var file = Path.Combine(_directory, assemblyName.Name);
            if (File.Exists(file + ".exe")) ReflectionOnlyLoadFromFileContent(file + ".exe");
            if (File.Exists(file + ".dll")) ReflectionOnlyLoadFromFileContent(file + ".dll");
            return Assembly.ReflectionOnlyLoad(args.Name);
        }

        private static Assembly ReflectionOnlyLoadFromFileContent(string file)
        {
            var bytes = File.ReadAllBytes(file);
            return Assembly.ReflectionOnlyLoad(bytes);
        }
    }
```

**Comment by bradwilson** [comment]
We can end up running user code during discovery. A reflection-only load will never work.

**Comment by TimLovellSmith** [comment]
>We can end up running user code during discovery. A reflection-only load will never work.

Hm. Personally I heretofore have never wanted to run user code during discovery. Although I can imagine that if I made that part of my mental toolbox I might want to do it more often... how would you feel about having a flag to opt-out of running user code during discovery?

**Comment by onovotny** [comment]
@TimLovellSmith doesn't work. Even `Fact` and `Theory` are user code in this context. The runner's don't bundle those, they come along with the xunit deps the test library brings in.

**Comment by frblondin** [comment]
> @TimLovellSmith doesn't work. Even Fact and Theory are user code in this context. The runner's don't bundle those, they come along with the xunit deps the test library brings in.

I'm not sure to understand what you are referring to. Were you meaning that getting the AttributeInfo needs the main AppDomain to load the specific version of the xunit assembly being referenced by the test assembly? See my sample code above.

As a recap the discovery (from my standpoint) could be done using two ways:

1. By using separate AppDomains (as of today) but in this case don't use byref marshaling but serialization. I don't think I received any response for my proposal above where I suggested that we could have another implementation of the ITypeInfo, IMethodInfo... interfaces that would use type being serialized instead of being marshaled, This would still allow getting data theory values.
2. We could use one single AppDomain and do reflection using Mono.Cecil. This would prevent from getting data theory values.

**Comment by TimLovellSmith** [comment]
>Were you meaning that getting the AttributeInfo needs the main AppDomain to load the specific version of the xunit assembly being referenced by the test assembly? See my sample code above.

I think the problem is that a) xunit calls code in the attribute object to get information - and b) attribute types are an extensibility point, although I am less certain of that?

**Comment by bradwilson** [comment]
Both those statements are true. xUnit.net has tons of extensibility points, and often uses them itself.

The conversation was talking about Fact/Theory and why code needs to be executed. The answer is that Fact/Theory are not special; everything goes through a discoverer class which understands the specific attribute and how to generate test cases for it.

From https://github.com/xunit/xunit/blob/b63031dc8698f9aaa94b006db4ccf20fde70f056/src/xunit.core/FactAttribute.cs:

```csharp
[XunitTestCaseDiscoverer("Xunit.Sdk.FactDiscoverer", "xunit.execution.{Platform}")]
[AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
public class FactAttribute : Attribute
```

That `[XunitTestCaseDiscoverer(...)]` attribute is used during discovery. We instantiate that class (in this case, the [`FactDiscoverer`](https://github.com/xunit/xunit/blob/b63031dc8698f9aaa94b006db4ccf20fde70f056/src/xunit.execution/Sdk/Frameworks/FactDiscoverer.cs)) and it does the work of generating test cases:

```csharp
public virtual IEnumerable<IXunitTestCase> Discover(ITestFrameworkDiscoveryOptions discoveryOptions, ITestMethod testMethod, IAttributeInfo factAttribute)
{
    var testCase =
        testMethod.Method.GetParameters().Any()
            ? new ExecutionErrorTestCase(diagnosticMessageSink, discoveryOptions.MethodDisplayOrDefault(), testMethod, "[Fact] methods are not allowed to have parameters. Did you mean to use [Theory]?")
            : CreateTestCase(discoveryOptions, testMethod, factAttribute);

    return new[] { testCase };
}
```

None of this is possible with a reflection-only assembly load.

**Comment by bradwilson** [comment]
Closed for age.
